core/templates/services/audio-player.service.ts

```diff
@@ -16,162 +16,219 @@
+          this._loadingTrack = false;
+          this._lastPauseOrSeekPos = 0;
+          this.play();
+          successCallback();
         });
-      },
-      play: function() {
-        _play();
-      },
-      pause: function() {
-        _pause();
-      },
-      stop: function() {
-        _stop();
-      },
-      rewind: function(seconds) {
-        _rewind(seconds);
-      },
-      forward: function(seconds) {
-        _forward(seconds);
-      },
-      getCurrentTime: function() {
-        if (_currentTrack) {
-          return Math.round(_currentTrack.currentTime);
-        } else {
-          return 0;
-        }
-      },
-      getAudioDuration: function() {
-        if (_currentTrack && _currentTrack.audio) {
-          return Math.round(_currentTrack.audio.duration);
-        } else {
-          return 0;
-        }
-      },
-      getProgress: function() {
-        if (!_currentTrack) {
-          return 0;
-        }
-        return _currentTrack.progress;
-      },
-      setProgress: function(progress) {
-        if (_currentTrack) {
-          _currentTrack.progress = progress;
-        }
-      },
-      isPlaying: function() {
-        return Boolean(_currentTrack && !_currentTrack.paused);
-      },
-      isTrackLoaded: function() {
-        return Boolean(_currentTrack);
-      },
-      clear: function() {
-        _currentTrack = null;
-        _currentTrackFilename = null;
-      },
-      get onAutoplayAudio() {
-        return _autoplayAudioEventEmitter;
-      }
-    };
-  }
-]);
+        this._currentTrack.on('end', () => {
+          this._currentTrack = null;
+          this._currentTrackFilename = null;
+          this._lastPauseOrSeekPos = null;
+          this.audioTranslationManagerService.clearSecondaryAudioTranslations();
+        });
+      }, (e) => errorCallback(e)
+    );
+  }
+
+  async loadAsync(filename: string): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this._loadAsync(filename, resolve, reject);
+    });
+  }
+
+  play(): void {
+    if (this.isPlaying()) {
+      return;
+    }
+    if (this._currentTrack !== null) {
+      this._currentTrack.seek(this._lastPauseOrSeekPos);
+    }
+    this.ngZone.runOutsideAngular(() => {
+      interval(500).pipe(takeUntil(
+        this._stopIntervalSubject)).subscribe(() => {
+        this.ngZone.run(() => {
+          this._updateViewEventEmitter.emit();
+        });
+      });
+    });
+    this._currentTrack.play();
+  }
+
+  pause(): void {
+    if (!this.isPlaying()) {
+      return;
+    }
+    this._lastPauseOrSeekPos = this.getCurrentTime();
+    this._currentTrack.pause();
+    this._stopIntervalSubject.next();
+  }
+
+  stop(): void {
+    if (!this._currentTrack) {
+      return;
+    }
+    this._lastPauseOrSeekPos = 0;
+    this._currentTrack.stop();
+    this._stopIntervalSubject.next();
+    this._currentTrackFilename = null;
+    this._currentTrack = null;
```

Comment:
Are we expecting to clear the audio when someone stops the audio (not pause), assume they replay it are we expecting to redownload the same audio?

Also, this looks similar to clear() function, if yes can we use that?

Reply:
This shouldn't be here. Ideally, The assets service caches these calls, so it is not re-downloaded. The stop function pauses the audio and sets the time to zero, the clear function removes the track.